pkg/exchange/ftx/stream.go

```diff
@@ -102,15 +116,77 @@ func (s *Stream) SetPublicOnly() {
 	atomic.StoreInt32(&s.publicOnly, 1)
 }
 
-func (s *Stream) Subscribe(channel types.Channel, symbol string, _ types.SubscribeOptions) {
-	if channel != types.BookChannel {
-		panic("only support book channel now")
+func (s *Stream) Subscribe(channel types.Channel, symbol string, option types.SubscribeOptions) {
+	if channel == types.BookChannel {
+		s.addSubscription(websocketRequest{
+			Operation: subscribe,
+			Channel:   orderBookChannel,
+			Market:    toLocalSymbol(TrimUpperString(symbol)),
+		})
+
+	} else if channel == types.KLineChannel {
+		// FTX does not support kline channel, do polling
+		interval := types.Interval(option.Interval)
+		ks := klineSubscription{symbol: symbol, interval: interval}
+		s.klineSubscriptions = append(s.klineSubscriptions, ks)
+	} else {
+		panic("only support book/kline channel now")
 	}
-	s.addSubscription(websocketRequest{
-		Operation: subscribe,
-		Channel:   orderBookChannel,
-		Market:    toLocalSymbol(TrimUpperString(symbol)),
```

Comment:
I wonder why can't we use the above method call to add subscription?

Reply:
The definition of websocketRequest is:
`type websocketRequest struct {
	Operation operation `json:"op"`

	// {'op': 'subscribe', 'channel': 'trades', 'market': 'BTC-PERP'}
	Channel channel `json:"channel,omitempty"`
	Market  string  `json:"market,omitempty"`

	Login loginArgs `json:"args,omitempty"`
}`

I suppose that It should be mapped to the json format of request to websocket. Since the kline is not supported by FTX, I decided not to modify the structure.

Or, the structure can be:

`type websocketRequest struct {
	Operation operation `json:"op"`

	// {'op': 'subscribe', 'channel': 'trades', 'market': 'BTC-PERP'}
	Channel channel `json:"channel,omitempty"`
	Market  string  `json:"market,omitempty"`

	Login loginArgs `json:"args,omitempty"`
+.     Interval types.Interval	
}`

Reply:
oh you're right, then I have no question :D